File: algorithms/medium/zigzag_conversion.cpp
class Solution {
public:
  string convert(string s, int numRows) {
    if (s.size() == 1 || s.size() <= numRows) {
      return s;
    }

    vector<string> arr(s.size());
    int i = 0;
    int step = 1;
    for (auto c : s) {
      arr[i] += c;
      if (i == 0) {
        step = 1;
      } else if (i == numRows - 1) {
        step = -1;
      }
      i += step;
    }
    stringstream res;
    copy(arr.begin(), arr.end(), ostream_iterator<string>(res, ""));
    return res.str();
  }
};

File: algorithms/medium/zigzag_conversion.ex
defmodule Solution do
  @spec convert(s :: String.t(), num_rows :: integer) :: String.t()
  def convert(s, num_rows) do
    len = String.length(s)

    if len == 1 || len <= num_rows do
      s
    else
      s
      |> String.graphemes()
      |> Enum.reduce({%{}, 0, 1}, fn c, {mp, i, step} ->
        step =
          cond do
            i == 0 -> 1
            i == num_rows - 1 -> -1
            true -> step
          end

        {Map.update(mp, i, [c], &[c | &1]), i + step, step}
      end)
      |> elem(0)
      |> Enum.sort_by(fn {k, v} -> k end)
      |> Enum.map(fn {k, v} -> Enum.reverse(v) end)
      |> Enum.join()
    end
  end
end
